[BE] [team-1 tree & 이노] 모든 api 생성

eNoLJ opened the pull request:
안녕하세요 team-1의 백엔드를 맡고 있는 tree와 이노입니다.
지난 pr과 다른 점은

1. 현재 사용할 api는 전부 만들었습니다.
timeline은 get 요청 외에 work의 save, update, remove, move가 발생할 때마다 기록되도록 하기위해 WorkService에 대부분의 로직이 생성되어있습니다. 이 부분에서 궁금한 사항이 생겼는데 이런식으로 timeline이 work에 강한 의존성이 보일 때 좋은 처리 방법이 있는지 궁금합니다.

2. WorkService 대부분의 메소드를 보면 1번 질문에 연장선으로 하나의 메소드 안에서 work의 저장과 timeline의 저장이 같이 실행이 되고있습니다. 이렇게 구현하는것이 잘못된 것인지 아니면 transaction 어노테이션을 사용하면 될까요??

3. utils 패키지에 TimelineContent 클래스가 있습니다. 하는 일은 timeline의 컨텐츠를 생성해주는 로직이 구현되어 있습니다. 이 역할을 현재처럼 TimelineContent 클래스를 따로 만들어 하는 것이 좋을까요?? 아니면 service나 domain 쪽에서 처리를 해주는 것이 좋을까요??

위에 질문들은 프로젝트 진행 시 많은 고민을 한 내용이기 때문에 꼭 답변 부탁드리겠습니다.😭
이번 리뷰도 잘부탁드려요!!

api문서 : https://app.gitbook.com/@todolist-2/s/todo-list/

zzisun committed:
Add: TO-DO LIST layout

torch-ray committed:
Add: Title StackView

torch-ray committed:
Feat: LoginViewController

torch-ray committed:
Fix: TaskCountLabel CornerRadius

zzisun committed:
Feat: containerView

zzisun committed:
Feat: containerView

zzisun committed:
Merge branch 'feature/iOS/todoList' of https://github.com/eNoLJ/todo-list into feature/iOS/todoList

zzisun committed:
Feat: addTaskButton

torch-ray committed:
Fix: CardViewController Constraint

zzisun committed:
Fix: View and Button CornerRadius

torch-ray committed:
Feat: implementation of addTaskbutton && CancelButton's flow

zzisun committed:
Feat: add new taskCard Action

torch-ray committed:
style: Delete Annotation and Add //MARK: description

torch-ray committed:
feat: Add TableViewCell Update Function

torch-ray committed:
style: Change BackgroundColor and Stackview Layout

zzisun committed:
Feat: add Buggit add -Agit add -A

torch-ray committed:
Feat: Add ViewController for Each Container View

torch-ray committed:
remove .DS_Store

zzisun committed:
Feat: Initial View

zzisun committed:
Feat: .gitignore

torch-ray committed:
Feat: TaskViewController Recycle Logic implementation

zzisun committed:
Feat: .gitignore

zzisun committed:
Feat: ino ga sikineondaro

torch-ray committed:
Style: Delete Annotation && ReplaceHardCoding with enum Type

torch-ray committed:
Style: Change BackgroundColor and Configure of TaskCountLabel

torch-ray committed:
Style: Add ButtonView

torch-ray committed:
feat: Add TaskCard Struct

torch-ray committed:
Feat: Add TaskCards && TaskCardsManager Class

torch-ray committed:
Sytle: Connect Title && Content TextField to NewTaskController

zzisun committed:
Style: NewTaskViewController & image

torch-ray committed:
Feat: Add TitleList EnumType

torch-ray committed:
Update README.md

torch-ray committed:
Update README.md

zzisun committed:
Update README.md

zzisun committed:
Update README.md

torch-ray committed:
Feat: Add NotificationCenter and Override Prepare Method to send status(id) to NewTaskViewController

torch-ray committed:
Feat: Deleate Annotation && Unused Method

torch-ray committed:
Feat: Delete Annotation

zzisun committed:
Style: TaskCell

eNoLJ committed:
Fix : session에서 User를 제대로 꺼내올 수 없는 문제 해결

session에 ResponseLoginUserDto가 아닌 User를 넘겨주도록 수정, session이 없을 때 예외로 사용할 IllegalUserAccessException 클래스 생성

eNoLJ committed:
Feat : TimelineController 클래스에 showTimeline 메소드 수정

기존에 정적으로 timeline의 더미 데이터만 반환하던 로직에서 동적으로 sessionUser를 확인해 반환하도록 수정

eNoLJ committed:
Feat : TimelineController 클래스에 showTimeline 메소드 수정

ResponseTimelineDto 클래스를 만들어 showTimeline 메소드로 요청이 들어왔을 때 dto로 변환하여 반환하도록 수정

eNoLJ committed:
Feat : WorkController 클래스에 showWorks 메소드 수정

ResponseWorkDto 클래스를 만들어 showWorks 메소드로 요청이 들어왔을 때 dto로 변환하여 반환하도록 수정

eNoLJ committed:
Feat : EntityRelatedException 클래스 생성

기존에 있던 EntityNotCreateException, EntityNotFoundException, EntityNotChangeException 클래스를 하나로 관리하기 위해 EntityRelatedException 클래스 생성

eNoLJ committed:
Feat : UserController 클래스에 signIn 메소드에서 ResponseUserDto를 반환하도록 수정

회원가입 요청에 있어서 또한 ResponseUserDto 객체를 전달하기 위해 ResponseCreateUserDto 클래스를 ResponseUserDto 클래스로 변경

eNoLJ committed:
Feat : WorkController 클래스에 work 관련 요청에 대한 메소드 생성

WorkController 클래스에 crud 기능을 제공하기 위해 createWork, updateWork, deleteWork, moveWork 메소드 생성, 이에 관련 service 로직 생성, RequestCreateWorkDto, RequestUpdateWorkDto 생성

eNoLJ committed:
Feat : RequestMoveWorkDto 클래스 생성

WorkController 클래스의 moveWork 메소드에서 RequestBody를 RequestMoveWorkDto 클래스를 이용하여 받도록 수정

eNoLJ committed:
Feat : Exception의 Response code 변경

EntityRelatedException과 UserAccountException이 발생 했을 때 좀 더 적절한 상태 코드를 보내도록 수정

choitree committed:
Feat : Timeline 생성 구현

Work가 생성,수정,삭제,이동시 Timeline이 생성되도록 구현

eNoLJ committed:
Refactor : WorkService 클래스의 timeline 로직 중복 제거

choitree committed:
Refactor : WorkService, TimelineDescription 클래스 중복 제거 및 리팩토링

zzisun committed:
Feat: Make iOS Folder

zzisun committed:
#5 Feat: Json encoding & decoding

zzisun committed:
Remove .DS_Store

zzisun committed:
Delete TodoApp/TodoApp.xcodeproj/project.xcworkspace directory

torch-ray committed:
#5 Feat: Delete test code

eNoLJ committed:
Refactor : TimelineDescription 클래스 수정

코드에 가독성을 높이기 위해 리팩토링

eNoLJ committed:
Style : build.gradle 파일의 의존성 정렬

eNoLJ committed:
Refactor : 불필요한 어노테이션 삭제

eNoLJ committed:
Refactor : WorkService 클래스에 불필요한 멤버변수 삭제

torch-ray committed:
#5 Feat: Store ParsingData From Network

eNoLJ committed:
Merge branch 'feature/BE/work' of https://github.com/eNoLJ/todo-list into feature/BE/work

eNoLJ committed:
Feat : .gitignore 추가

torch-ray committed:
#5 Feat: Add drawTaskView Method

eNoLJ committed:
Merge branch 'team-1' of https://github.com/eNoLJ/todo-list into team-1

eNoLJ committed:
Merge branch 'team-1' of https://github.com/eNoLJ/todo-list into feature/BE/work

choitree committed:
Refactor : 변수명 author와 createdTime을 authorId, createdAt으로 변경

Database에는 created_at, author_id로 변경

zzisun committed:
#5 Feat: send removed data

torch-ray committed:
#5 Feat: Add TaskCard Method

eNoLJ committed:
Merge pull request #14 from eNoLJ/feature/BE/work

모든 api 생성

torch-ray committed:
#5 Feat: Add insertedDataPost Method

torch-ray committed:
TEST

choitree committed:
#17 Refactor : build.gradle에 lombok 의존성 추가

choitree committed:
[#17] Refactor : User, Timeline, Work 클래스에 @Getter, @Setter, @ToString 어노테이션 추가

choitree committed:
[#17] Refactor : lombok @Getter, @Setter, @ToString 어노테이션 추가

torch-ray committed:
#5 Refactor: SpellingError

choitree committed:
[#17] Refactor : description 컬럼명을 content로 변경

choitree committed:
[#17] Refactor : TimelineDesption 클래스명을 TimelineContent로 변경

choitree committed:
[#17] Refactor : lombok 관련하여 xml 파일 내용 추가

zzisun committed:
#5 Feat: sync data with DB

torch-ray committed:
Merge pull request #18 from eNoLJ/feature/iOS/network

Feature/i os/network

torch-ray committed:
Merge pull request #19 from eNoLJ/feature/iOS/todoList

Feature/i os/todo list

eNoLJ committed:
Test : 빌드를 위해 테스트 케이스 주석 처리

eNoLJ committed:
Feat : spring.datasource.url 변경

배포되어 있는 mysql을 사용하던 부분을 로컬의 mysql을 사용하도록 변경

eNoLJ committed:
[#17] Refactor : lombok의 builder를 활용하여 리팩토링

User, work, Timeline, 관련 dto 클래스의 생성자를 builder를 활용하여 리팩토링

eNoLJ committed:
Feat : gitignore 파일 추가

eNoLJ committed:
Merge pull request #21 from eNoLJ/feature/BE/refactor

롬복을 사용하여 리팩토링

zzisun committed:
Chore: merge with BE

ksundong commented on .gitignore:
```diff
@@ -106,3 +106,9 @@ fabric.properties
 .idea/codestream.xml
 
 # End of https://www.toptal.com/developers/gitignore/api/intellij
+.idea/.gitignore
+.idea/compiler.xml
+.idea/jarRepositories.xml
+.idea/modules.xml
+.idea/todo-list.iml
+.idea/vcs.xml
```
이미 버전관리되고 있는 파일은 단순히 `gitignore`에 추가한다고, 무시되지 않습니다.
이미 버전관리되고 있는 부분을 제거해주셔야 합니다.
`git rm --cached`에 대해서 학습하시고, 제거해보세요.

eNoLJ replied:
현재 말씀하신 파일들은 올라가있지 않습니다. 최상위 폴더인 todo-list 폴더의 .idea를 올리고 싶지 않아서 추가를 해줬습니다. 그리고 바로 하위 폴더인 BE에는 .idea가 올라가 있습니다.

ios와 be가 폴더를 나눠 작업을 하다보니 최상위 BE 폴더뿐만 아니라 최상위 폴더에도 .idea가 생겨서 그 부분만 없애도록 했습니다!!

ksundong replied:
말씀하신 것 이외에도 버전관리 되지 않아야 하는 파일들이 버전관리 되고있습니다.
확인해주시고, 다시 코멘트 남겨주세요.

eNoLJ replied:
@ksundong BE/.idea도 추가로 제외했습니다.

ksundong commented on BE/build.gradle:
```diff
@@ -13,11 +13,13 @@ repositories {
+	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
 	developmentOnly 'org.springframework.boot:spring-boot-devtools'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	implementation 'org.projectlombok:lombok'
+	annotationProcessor 'org.projectlombok:lombok'
```
저는 롬복 사용에 있어서 학습시점에 도입은 아직까진 좋지 않다고 생각합니다.
좋은데, 아직 제대로 쓰기에는 많은 어려움이 있을것이라고 생각돼요.
`@Builder` 같이 정말 verbose 해지는 것이 아니라면, 의존성은 넣어두되, 간단한 코드는 직접 작성해서 사용하는 편을 더 추천드리겠습니다.

eNoLJ replied:
@ksundong 안녕하세요 디온!! 하나하나 자세히 리뷰를 해주셔서 먼저 감사하단 말씀 드리고 싶습니다. 롬복에 대해 지적을 해주셨는데 저희도 롬복을 사용하지 않았다가 전에 리뷰어 분이 한번 사용해보면 어떻겠냐하셔서 사용해 봤습니다. 사용할 땐 코드도 간결해지고 깔끔한 느낌이 들어 좋다고 사용했는데 디온의 리뷰를 보고 무슨 문제가 생길까 찾아봤습니다. 무한루프에 빠질 수 있다는 점, 코드가 생략되어 가독성이 낮아진다는 점 등을 찾아볼 수 있었습니다.

저희도 이번에 처음 롬복을 직접 사용하면서 배우고 디온 덕에 문제점을 학습할 수 있었습니다. 이번 프로젝트에서는 롬복에 대해 경험하고 학습하는 기회로 봐주시면 감사하겠습니다. 다음 프로젝트부터는 사용하지 않도록 하겠습니다.

ksundong replied:
네 롬복은 코드를 간단히 작성해 줄 수 있게 해주는 대신에, 무엇이 작성되어야 하는지 정확히 판단하기 힘든 초보시절엔 비추천 드리는 의존성 중 하나입니다.

1. 잘못 사용한다면, 무한 루프에 빠질 수도 있겠네요.
2. 코드가 생략되는것이 왜 가독성이 떨어질까요? 저는 롬복 사용 자체는 가독성을 획기적으로 높여준다고 생각해요.

쓰셔도 되는데, 이런 라이브러리를 사용하실 때에는 항상 많은 검색을 해보시고 사용해보시길 추천드립니다.
https://cheese10yun.github.io/lombok/
이 글이 도움이 되실 것 같네요.

eNoLJ replied:
글 참고하여 좀 더 공부해보도록 하겠습니다. 감사합니다!!

ksundong commented on BE/src/main/java/com/example/todolist/domain/timeline/Timeline.java:
```diff
@@ -1,52 +1,24 @@
 
 import java.time.LocalDateTime;
 
+@ToString
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
롬복은 이렇게 사용하시면 굉장히 큰 난관을 겪으실 수 있습니다.

choitree replied:
안녕하세요 디온 트리입니다.
이번 리뷰도 세세하게 정말정말 감사합니다. 
롬복 관련하여 리뷰 주신 내용으로 롬복의 단점 등을 키워드로 학습해보았는데, 
디온이 말씀해주시는 난관이 무한루프에 빠질 수 있다는 점인가요?~ 아니면 다른 문제가 발생할 수 있는 가능성이 있나요??

ksundong replied:
`@Getter`를 제외하면 다 조심하셔야 합니다.
컴파일된 코드를 보시고, 의도하신대로 만들어졌는지도 확인해보세요.

ksundong commented on BE/src/main/java/com/example/todolist/domain/timeline/TimelineRepository.java:
```diff
@@ -1,13 +1,11 @@
 public interface TimelineRepository extends CrudRepository<Timeline, Long> {
 
-    List<Timeline> findAllByAuthor(Long author);
+    List<Timeline> findAllByAuthorId(Long author);
```
매개변수명도 변경되면 좋을 것 같네요!

choitree replied:
넵!

ksundong commented on BE/src/main/java/com/example/todolist/domain/work/Work.java:
```diff
@@ -1,68 +1,54 @@
     private int status;
 
-    private Long author;
+    private Long authorId;
```
`User` 타입이면 안되는 이유가 있나요?

eNoLJ replied:
안될 이유 없습니다!! 하지 못한거에요...ㅎㅎ jdbc는 jpa와는 달리 알아서 해주지 않아서 생략하고 진행했습니다. 좀 더 찾아서 해보고 다시 코멘트 남기겠습니다.

ksundong replied:
넵 화이팅입니다.

eNoLJ replied:
@ksundong 다른 리뷰는 현재 거의 다 반영을 한것 같은데 이 부분이 해결이 안되고 있습니다. jdbc에서 manyToOne을 사용한 방법이나 여러가지를 찾아봐도 찾을 수가 없네요... 혹시 검색 키워드나 방향을 제시해주실 수 있을까요??

ksundong replied:
제가 잘못 리뷰했어요 ㅠㅠ JPA랑 착각했네요~

ksundong commented on BE/src/main/java/com/example/todolist/domain/work/Work.java:
```diff
@@ -1,68 +1,54 @@
-
-    public String getTitle() {
-        return title;
+    public void save(User sessionUser) {
+        this.authorId = sessionUser.getId();
```
저장한다는 의미와는 사뭇 다른 것 같아요.

eNoLJ replied:
save 메소드 이름을 saveAuthorId로 변경 했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/domain/work/Work.java:
```diff
@@ -1,68 +1,54 @@
-        return createdTime;
+    public void delete() {
+        this.status = 0;
     }
```
status가 의미하는 바가 무엇인가요?
ENUM을 사용해보는건 어떨까요?

eNoLJ replied:
현재 request와 response에는 ios쪽과 int로 소통하는걸로 정해서 디온의 리뷰를 보고 work entity 클래스의 status만 Status enum을 만들어 사용하도록 수정했습니다.

ksundong replied:
그냥 반환할 때만 Enum에서 int를 꺼내줘도 될 것 같고, int로 소통한다는건 제 생각엔 굉장히 많이 불편할 것 같네요~

eNoLJ replied:
아침에 ios쪽과 상의한 결과 현재로서 통신 부분에 있어 enum으로 하기에는 시간이 촉박하다 판단되어 이번 프로젝트는 int로 진행하도록 하겠습니다. 다음 프로젝트 부터는 처음 기획할 때부터 enum을 염두에 두고 진행해보도록 하겠습니다. 감사합니다.

ksundong replied:
서버쪽에서만 Enum을 사용해도 됩니다~ 생각을 너무 한정짓지말아주세용~

ksundong commented on BE/src/main/java/com/example/todolist/exception/EntityRelatedException.java:
```diff
@@ -0,0 +1,13 @@
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class EntityRelatedException extends RuntimeException {
+
+    public EntityRelatedException(ErrorMessage errorMessage) {
+        super(errorMessage.getErrorMessage());
+    }
```
조금 확장성이 떨어져보이네요.
기본 메시지를 이곳에서 직접 반환해주는 것도 고려해보세요.

ksundong replied:
String으로도 받아서 반환하는 것도 생각해주세요.

eNoLJ replied:
@ksundong 디온 말씀의 뜻이 매개변수로 string으로 받도록 하라는 말씀이신가요?? 그리고 저 exception을 호출할 때 메세지 자체를 넣는 식으로 변경을 하구요?? 혹시 맞다면 그렇게 하는게 더 좋은 이유가 무엇인가요?? 오히려 호출하는 곳에서 enum의 string을 빼는 코드가 추가되어 더 번거로울 것 같아서 위처럼 만들었던건데 한번 더 설명해주시면 감사하겠습니다!!

ksundong replied:
1. 기본 메시지를 반환하는 예외 생성자를 만들어주세요.
2. ErrorMessage를 인자로 받는것 이외에 생성자 오버로딩을 이용해서 String도 받도록 해주세요.

예외 메시지는 동적으로 변하는 메시지가 작성되어야 할 필요가 있을 수도 있습니다. 그 밖에도 Enum으로 대응하기 힘든 상황도 있을거에요. 예외 메시지 자체를 Enum으로 받는건 좋은 아이디어지만, Enum으로만 받는건 별로 좋지 못하다고 생각됩니다.

호출하는 쪽에서 enum의 string을 빼는 경우를 말씀드린게 아닙니다.

eNoLJ replied:
현재 만들어져 있는 exception 3개에 모두 String을 매개변수로 받는 생성자를 오버라이딩 했습니다.

ksundong replied:
오버라이딩 vs 오버로딩에 대해서도 한 번 공부해보세요~|

ksundong commented on BE/src/main/java/com/example/todolist/service/TimelineService.java:
```diff
@@ -0,0 +1,26 @@
+        return timelineRepository.findAllByAuthorId(sessionUser.getId()).stream()
+                .map(timeline -> new ResponseTimelineDto(timeline, sessionUser))
+                .collect(Collectors.toList());
+    }
```
stream의 활용 아주 좋습니다.

ksundong commented on BE/src/main/java/com/example/todolist/service/WorkService.java:
```diff
@@ -0,0 +1,88 @@
+        Work work = workDto.toEntity();
+        work.save(sessionUser);
+        Work saveWork = workRepository.save(work);
+        saveTimeline(work, makeTimelineContent(work, "save"));
```
save는 상수로 추출해주는 편이 좋을 것 같아요~

ksundong commented on BE/src/main/java/com/example/todolist/service/WorkService.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public ResponseWorkDto update(Long id, RequestUpdateWorkDto workDto, User sessionUser) {
+        Work work = verifyWork(id, sessionUser);
```
검증한다는 의미와는 다르게 값을 반환하네요!

eNoLJ replied:
verifyWork 메소드의 이름을 getVerifiedWork로 변경했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/service/WorkService.java:
```diff
@@ -0,0 +1,88 @@
+
+    public ResponseWorkDto move(Long id, RequestMoveWorkDto workDto, User sessionUser) {
+        Work work = verifyWork(id, sessionUser);
+        work.move(workDto.toEntity());
```
move는 뭔가 향해야 할 대상을 받아야 할 것 같은데, 그냥 전체 갱신을 시키는 느낌이네요.

eNoLJ replied:
어디로 이동을 해야하는지 알 수 있도록 하라는 말씀으로 이해를 하여
`work.move(workDto.toEntity().getStatus())`로 변경, work의 move 메소드의 매개변수를 int로 변경했습니다.
추후 타입은 enum으로 변경할 수 있도록 고민해 보겠습니다.

ksundong commented on BE/src/main/java/com/example/todolist/service/WorkService.java:
```diff
@@ -0,0 +1,88 @@
+        return buildResponseWorkDto(work, sessionUser);
+    }
+
+    public void saveTimeline(Work work, String content) {
+        timelineRepository.save(Timeline.builder()
+                .content(content)
+                .authorId(work.getAuthorId())
+                .createdAt(LocalDateTime.now())
+                .build());
+    }
```
WorkService인데 Timeline을 저장하는 것 뭔가 이상하지 않나요?

eNoLJ replied:
@ksundong pr 보낼 때 1번 질문이 이러한 행동에 대해 말씀드렸던 부분입니다. 이런식으로 WorkService에 timeline에 대한 로직이 대부분 들어가 있어서 만약 좋은 코드가 아니라면 어떤식으로 분리를 해야하는지 궁금했어요. 만약 위에 말씀하신 save 코드를 TimelineService에 만든다면 service에서 service를 사용하게 되는데 이 부분도 적절하지 않아보였어요.

이 부분 지적해주실 때 생각하신 방법이 있다면 조언 부탁드리겠습니다!!

ksundong replied:
왜 Service에서 다른 Service를 이용하면 안되나요?

eNoLJ replied:
@ksundong 전에 그룹에서 얘기가 나왔었는데 지양하는 분위기였어서 "아 서비스에서 서비스를 사용하면 어색하구나"싶은 막연한 생각만 갖고 있었습니다. 근데 디온이 말씀하신거보면 그건 또 아닌가보네요?? 그래서 TimelineService에 로직을 생성하고 WorkService 수정했습니다.

ksundong replied:
저는 적어도 Service에서 여러 Repository 를 호출하는 구조보다는 훨씬 좋은 구조라고 생각합니다.

ksundong replied:
https://www.inflearn.com/questions/21703

ksundong replied:
다만 이 구조를 택할경우 주의할 점이 사이클이 생기는 것을 경계해야 한다는 점입니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/TimelineController.java:
```diff
@@ -1,31 +1,31 @@
 public class TimelineController {
 
-    public final TimelineRepository timelineRepository;
+    private final Logger logger = LoggerFactory.getLogger(TimelineController.class);
+    public final TimelineService timelineService;
```
보통 java에서는 public이 private보다 위에 위치하도록 하는 것이 일반적입니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/TimelineController.java:
```diff
@@ -1,31 +1,31 @@
-            return new ArrayList<>();
-        }
-        return timelineRepository.findAllByAuthor(1L);
+    public List<ResponseTimelineDto> showTimeline(HttpSession session) {
```
List형태로 반환하면
```json
[
  {},
  {}
]
```

형태가 될겁니다.
만약, 개수를 반환해야한다는 조건이 들어가면, API의 breaking change가 발생하게됩니다.
이를 방지하기 위해서
```json
{
  "timelines": []
}
```
와 같은 형태로 반환하는 것을 고려해주세요.

eNoLJ replied:
넵 현재 코드는 List형태로 넘겨줘서 개수와 같은 다른 정보가 추가적으로 필요할 때 디온의 말씀대로 변경할 필요성이 있습니다. 하지만 현재 프로젝트 시작 시 ios와 상의한 결과 리스트 외에 정보를 넘겨줄 필요성은 없을 것 같아서 간단하게 List를 넘겨주도록 하였습니다. 그리고 이 부분은 ios까지 변경할 필요성이 생겨서 프로젝트가 이틀 남은 현재로서는 우선순위가 낮다고 판단이 됩니다.
물론 디온의 말씀의 의도는 충분히 파악하고 있으니 다음부터는 꼭 염두에두고 작성하도록 하겠습니다!!!

ksundong commented on BE/src/main/java/com/example/todolist/web/TimelineController.java:
```diff
@@ -1,31 +1,31 @@
-        }
-        return timelineRepository.findAllByAuthor(1L);
+    public List<ResponseTimelineDto> showTimeline(HttpSession session) {
+        logger.info("타임라인 리스트 요청");
```
어떻게 해야 보다 도움이 되는 정보를 포함하는 로그를 남길 수 있을까요?

eNoLJ replied:
config 패키지에 RequestLoggingFilterConfig 클래스를 만들어 설정을 통해 request의 정보가 찍히도록 했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/UserController.java:
```diff
@@ -25,23 +26,23 @@ public UserController(UserService userService) {
 
     @PostMapping("/signIn")
-    public void signIn(@RequestBody RequestSignInUserDto requestUserDto) {
+    public ResponseUserDto signIn(@RequestBody RequestSignInUserDto requestUserDto) {
```
DTO는 Data Transfer Object의 약자로 `XXXDTO`의 형태로 쓰는것이 일반적입니다.

eNoLJ replied:
말씀대로 DTO로 변경 했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/UserController.java:
```diff
@@ -25,23 +26,23 @@ public UserController(UserService userService) {
     public void logout(HttpSession session) {
         logger.info("로그아웃 요청");
-        session.removeAttribute("sessionUser");
+        session.removeAttribute(USER_SESSION_KEY);
```
상수 처리 좋네요.

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/RequestCreateWorkDto.java:
```diff
@@ -0,0 +1,34 @@
+@Setter
+public class RequestCreateWorkDto {
+
+    private String title;
+
+    private String content;
+
+    private int status;
```
```suggestion
    private String title;
    private String content;
    private int status;
```

eNoLJ replied:
entity 클래스의 @id 멤버변수를 제외한 멤버변수와 dto 클래스의 멤버변수 모두 줄바꿈 제거 했습니다.

ksundong replied:
사실 뭐 남기셔도 상관없긴 합니다 ㅎㅎ

ksundong replied:
컨벤션을 정하시고 컨벤션 대로 코딩해주세요~

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/RequestCreateWorkDto.java:
```diff
@@ -0,0 +1,34 @@
+    private int status;
+
+    public Work toEntity() {
+        if (title == null || content == null || status == 0) {
+            throw new EntityRelatedException(ErrorMessage.ENTITY_NOT_CREATE);
+        }
```
따로 메서드로 추출해도 좋을 것 같아요~

eNoLJ replied:
해당 부분 `verifyDTO()` 메소드를 생성하여 분리했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/RequestLoginUserDto.java:
```diff
@@ -1,8 +1,13 @@
+import lombok.Setter;
+import lombok.ToString;
 
+@ToString(exclude = "password")
```
👍

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/RequestMoveWorkDto.java:
```diff
@@ -0,0 +1,25 @@
+    private int status;
+
+    public Work toEntity() {
+        if (status != 1 && status != 2 && status !=3) {
+            throw new EntityRelatedException(ErrorMessage.ENTITY_NOT_CREATE);
+        }
```
이 부분도 마찬가지로 추출해주시고, 과연 status가 이렇게 관리되어야 할지 고민해봐주세요.

eNoLJ replied:
ios 쪽에서 request로 받을때 int로 받아서 저 부분을 저렇게 했습니다. 현재 고쳐본다고하면 int를 Status로 바꿔서 Status를 비교하는 방식으로 해야할 것 같은데 그것보다 현재로서는 int를 비교하고 Status로 변경해주는게 낫다고 판단이 되네요. 혹시라도 내일 ios 쪽과 해당부분에 대해 얘기를 나누고 enum으로 받을 수 있게 된다면 수정해보도록 하겠습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/RequestUpdateWorkDto.java:
```diff
@@ -0,0 +1,28 @@
+
+@ToString
+@Setter
+public class RequestUpdateWorkDto {
```
UpdateWorkRequest는 어떨까요~?

eNoLJ replied:
말씀해주신대로 DTO 클래스의 이름 형식 바꿨습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/ResponseTimelineDto.java:
```diff
@@ -0,0 +1,32 @@
+
+@ToString
+@Getter
+
```
여기 빈라인은 제거해주시는게 좋을 것 같아요~

ksundong commented on BE/src/main/java/com/example/todolist/web/dto/ResponseTimelineDto.java:
```diff
@@ -0,0 +1,32 @@
+
+    private LocalDateTime createdAt;
+
+    private String author;
```
author가 String이 될 수도 있군요?

eNoLJ replied:
@ksundong response로 timeline을 넘겨줄 때 ios 쪽에서 author의 이름이 필요해서 저 부분만 변경해서 보내주고 있습니다.

ksundong replied:
그렇다면 authorName정도가 적당하지 않을까 싶네요~
이건 정하기 나름입니다~

eNoLJ replied:
참고하도록 하겠습니다!! 감사합니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/utils/TimelineContent.java:
```diff
@@ -0,0 +1,57 @@
+import com.example.todolist.web.dto.RequestMoveWorkDto;
+
+public class TimelineContent {
+    public static String makeTimelineContent(Work work, String method) {
+        String content = "";
+        content += getStatus(work.getStatus()) + "에 ";
+        content += verifyJongsung(work.getTitle(), "을", "를");
+
+        switch(method) {
+            case "save" :
+                return content + " 등록하였습니다.";
+            case "update" :
+                return content + " 수정하였습니다.";
+            case "delete" :
+                return content + " 삭제하였습니다.";
+            default :
+                return "";
+        }
+    }
```
String Builder를 써서 개선해봅시다.
그리고 문자열은 상수로 만들어주세요~

ksundong replied:
또한 switch-case 문을 사용하지 않으려면 어떻게 해야할 지도 고민해봐주세요!

eNoLJ replied:
String Builder를 사용하여 리팩토링 했습니다.
switch-case문은 좀 더 고민 후 수정하면 다시 한번 말씀드리도록 하겠습니다.

eNoLJ replied:
현재 Status enum과 Action enum을 활용하여 TimelineContent 클래스의 코드가 많이 간략해 졌습니다. switch-case문도 제거 했습니다.

ksundong commented on BE/src/main/java/com/example/todolist/web/utils/TimelineContent.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public static final String verifyJongsung(String title, String firstValue, String secondValue) {
+        char lastTitle = title.charAt(title.length() - 1);
+        //한글의 제일 처음과 끝의 범위밖일 경우는 오류
+        if (lastTitle < 0xAC00 || lastTitle > 0xD7A3) {
+           return title;
+        }
+        String seletedValue = (lastTitle - 0xAC00) % 28 > 0 ? firstValue : secondValue;
+
+        return title + seletedValue;
```
고생하셨네요.

ksundong commented on BE/src/test/java/com/example/todolist/TodoListApplicationTests.java:
```diff
@@ -6,8 +6,8 @@
-	@Test
-	void contextLoads() {
-	}
+//	@Test
+//	void contextLoads() {
+//	}
```
사용하지 않는다면 제거해주세요.

eNoLJ replied:
수동으로 빌드할 때 저 test 메소드가 문제가 생겨 주석처리를 했었는데 말씀대로 사용하지 않아 제거 했습니다!!

ksundong reviewed (changes requested):
안녕하세요! 리뷰어 Dion입니다! 벌써 두 번째 PR을 올려주시다니! 대단하십니다. 👍 

1. 애매한 질문이네요. 질문하고자 하는것이 무엇인가요? 비즈니스 로직이 의존하는 관계를 강제하는데, 이렇게 의존하는 것을 코드상에서 분리하는 것을 의도하는 걸까요?
2. 네, `@Transactional`을 붙여서 작업하는 것이 맞습니다. 기본적으로 `Service`에는 read only 속성을 주고, 나머지 트랜잭션이 필요한 부분에 모두 붙여주시는 것을 권장합니다.
3. 출력 로직이기 때문에 별도의 영역에 분리해서 관리하거나, 백엔드에서는 단순 데이터만 저장하고, 앞단에서 조립하는 방법도 있겠네요.

고생하셨습니다. 개선할 포인트가 보여 변경요청 드리니 수정 후에 리뷰요청 보내주세요!
궁금한 점이나, 도움이 필요한 부분, 이해가 되지 않는 부분, 토론하고 싶은 부분은 언제든 코멘트 남겨주세요~

ksundong commented:
또, API 문서를 작성하는 것은 매우 좋습니다.
하지만 제공해주신 API 문서만 보고 개발할 수 있을까요?
API 문서의 목적에 대해서 다시 한 번 생각해보셨으면 좋겠습니다.

zzisun committed:
Chore: remove test Todo folder

zzisun committed:
Chore: remove .DS_Store file in iOS Folder

eNoLJ commented:
@ksundong API문서의 목적은 클라이언트 측에서 사용할 때 어떤 정보를 넘겨주면 어떤 정보를 받는지 알려주는 용도로 사용된다고 생각하고 있습니다. 그런 의미에서 현재 작성된 API문서를 보고 개발을 할 수 있지 않나요?? 혹시 다른 목적이 또 있는건가요??

zzisun committed:
#5 #10 #11 Feat: longPress menuItem

zzisun committed:
Chore: fix storyboard unnecessary thing

zzisun committed:
#5 Feat: longPress menuItem delete function

choitree committed:
Refactor : Dto 클래스명을 DTO로 변경, 변수 줄바꿈 삭제

zzisun committed:
#5 Feat: longPress menuItem move to done function (with bug)

choitree committed:
Refactor : 변수 줄바꿈 삭제

choitree committed:
Feat : logging level 변경

choitree committed:
Feat : Request 작업 시 로그로 정보 확인하기 위해 config 추가

choitree committed:
Refactor : parameter 수정

choitree committed:
Refactor : DTO 클래스 변경에 따른 리턴 값 변경

choitree committed:
Refactor : 상수 추가, StringBuilder 추가하여 코드 리팩토링

choitree committed:
Refactor : DTO 클래스명 변경에 따른 리턴값 수정

choitree committed:
Refactor : Test 클래스에 사용하지 않는 메소드와 어노테이션 삭제

zzisun committed:
#5 Feat: longPress menuItem move to done function (with bug)

zzisun committed:
#5 Chore: add unpushed data

choitree committed:
Refactor : BE.iml 추가

choitree committed:
Refactor : entity의 null 체크하는 로직 verifyDTO 메소드로 추출

choitree committed:
Refactor : Dto로 되있는 parameter DTO로 수정

choitree committed:
Refactor : TimelineContent 클래스에 switch-case 문 사용 제거를 위해 Action Enum
클래스 생성 및 메소드, 파라메터 수정

eNoLJ committed:
Feat : schema.sql의 work 테이블에 status 컬럼 int 타입에서 varchar(64)로 변경

status를 enum으로 관리하기 위해 status의 타입을 int에서 varchar로 변경

eNoLJ committed:
Feat : Status enum 생성

Work entity의 status를 int가 아닌 enum으로 관리하기 위해 Status 이넘 생성

eNoLJ committed:
Refactor : build 메소드 이름 리팩토링

eNoLJ committed:
Refactor : Status enum을 활용하여 리팩토링

eNoLJ committed:
Chore : conflict 해결

eNoLJ committed:
Feat : Status enum에 comment 멤버변수 생성

TimelineContent 클래스에서 Status의 comment를 사용하기 위해 Status enum에 comment 생성

torch-ray committed:
#5 Chore: sync with team-1

ksundong commented:
해당 문서를 다른 팀 개발자 분에게 보여드리는 것도 좋은 방법이라고 생각됩니다.

![image](https://user-images.githubusercontent.com/38597469/114699630-84c4b200-9d5b-11eb-82e4-1b71a7438e99.png)

해당 이미지를 보면, 이 엔드포인트가 하는 일이 회원가입인데, 아래에는 이상한 설명(?)이 적혀있습니다.
Request는 모두 string인데, 모두 Optional 합니다. 거기다가 어떤 정보를 넣어줘야하는지도 모르겠습니다.
Response를 보아야 어떤 형태로 입력을 할 수 있는지 유추가 가능합니다. 그마저도 password가 가능한 형태는 문자열 말고는 어떤 정보가 없네요.

학습해가는 과정이고, 문서화의 중요성은 상대적으로 낮습니다만, 좋은 문서화는 어떻게 해야할지도 고민해보는 것도 좋을 것 같다고 생각됩니다~

torch-ray committed:
Test Ray

ksundong commented:
- Conflict 발생하는 부분 확인해주시고, 수정해주세요!
- gitignore 된 파일 제거 확인해주세요.
모두 완료되시면, 리뷰 요청 다시 보내주세요!!

ksundong commented:
API 문서를 작성하려는 시도는 아주 좋았어요!
앞으로 코쿼과정을 해나가면서 좀 더 좋은 API 문서에 대해서 계속해서 고민을 해보시는걸 추천드려요~

torch-ray committed:
#5 Feat: MoveToDoneView Update Function

torch-ray committed:
Merge pull request #23 from eNoLJ/rayTest

Feat: MoveToDoneView Update Function

torch-ray committed:
#5 Feat: Block move from done to done code

torch-ray committed:
#5 Chore: Sync Data

zzisun committed:
Merge pull request #25 from eNoLJ/dev-iOS

[iOS] first reviewed

torch-ray committed:
Create README.md

choitree committed:
Chore : conflict 해결

eNoLJ commented:
@ksundong API문서 만들 때 하나하나 무슨뜻인지 찾아보지 않고 작성을 했네요. 디온이 언급해주셔서 이제서야 의식이 됩니다. 찾아보고 보완하도록 하겠습니다. 그리고 하나하나 디테일하게 리뷰해주셔서 정말 감사합니다. 다음에도 제 이름이 보이면 잘 부탁드리겠습니다ㅎㅎ

choitree committed:
Style : 띄어쓰기 추가

choitree committed:
Chore : 충돌나서 생성된 파일 삭제

eNoLJ committed:
Chore : .gitignore 파일 추가

eNoLJ committed:
Merge branch 'dev-BE' of https://github.com/eNoLJ/todo-list into dev-BE

eNoLJ committed:
Feat : exception 클래스에 매개변수로 String errorMessage를 받아 보여줄 수 있도록 생성자 추가

eNoLJ committed:
Feat : TimelineService 클래스에 메소드 생성

Timeline에 관련 서비스를 WorkService에서 분리하기 위해 TimelineService 클래스에 editByAction, move, saveTimeline, makeTimelineContent, moveContent, verifyJongsung 메소드 생성
불필요한 TimelineContent 클래스 삭제
TimelineService 클래스 변경에 따라 관련 로직 수정

eNoLJ committed:
Feat : Application TimeZone 설정

배포 했을 때 한국 시간으로 mysql에 저장하기 위해 TimeZone 설정

eNoLJ committed:
Feat : Work 테이블에 updated_at 컬럼 생성

Work 테이블의 레코드를 updated_at 시간을 기준으로 정렬해 보여주기 위해 updated_at 컬럼 생성

eNoLJ committed:
Feat : Work 엔티티 클래스에 updatedAt 멤버변수 생성

Work 엔티티가 수정된 시간을 기록하기 위해 updatedAt 멤버변수 추가, 해당 로직 수정

choitree commented:
디온 API 에 대해서 회원가입에 적혀있는 이상한 설명은 삭제하고 알맞은 문서가 되도록 수정하겠습니다.

학습하다가 궁금한 점이, Body Parameter에 parameter마다 Optional과 required를 선택하도록 되어 있는데 
제가 생각하기로는 required는 전달하지 않으면 안되는 parameter고
Optional은 null로 전달되어도 되는 값일 때 선택하면 된다고 이해했는데, 이에 대해서 알맞게 이해한 것이 맞는지 확인 부탁드립니다.

데이터의 타입 뿐 아니라 어떤 값이 전달되어야 하는지 예시에 대해서도 추가하고
메소드에 대해서도 간략하게 설명할 수 있도록 하겠습니다.

ksundong commented:
> 디온 API 에 대해서 회원가입에 적혀있는 이상한 설명은 삭제하고 알맞은 문서가 되도록 수정하겠습니다.
> 
> 학습하다가 궁금한 점이, Body Parameter에 parameter마다 Optional과 required를 선택하도록 되어 있는데
> 제가 생각하기로는 required는 전달하지 않으면 안되는 parameter고
> Optional은 null로 전달되어도 되는 값일 때 선택하면 된다고 이해했는데, 이에 대해서 알맞게 이해한 것이 맞는지 확인 부탁드립니다.
> 
> 데이터의 타입 뿐 아니라 어떤 값이 전달되어야 하는지 예시에 대해서도 추가하고
> 메소드에 대해서도 간략하게 설명할 수 있도록 하겠습니다.

네 Required는 필수 라는 의미고, Optional은 선택적으로 보내줘도 된다는 의미입니다.
회원가입시에 필수정보라고 볼 수 있는 아이디, 비밀번호, 이메일, 이름은 당연히 필수적이겠죠? 경우에 따라서는 프로필 사진을 선택적으로 받을 수 있다. 정도의 기능도 구현해볼 수 있을겁니다.

네 여유가 되시면 문서에도 신경을 써주시고, 없으시면 너무 세세하게 정리해주실 필요는 없다고 생각해요~

ksundong commented:
머지를 하려고 했더니 Swift 파일이 포함되어있네요 ㅠㅠ 어떻게 된 일일까요?

eNoLJ commented:
> 머지를 하려고 했더니 Swift 파일이 포함되어있네요 ㅠㅠ 어떻게 된 일일까요?

@ksundong 현재 밖이라 다른 리뷰에 관해서는 추후 답변을 달겠습니다.

team-1브랜치는 ios와 be가 통합적으로 되어있는 구조입니다. 어제 ios쪽에 머지가 되어서 그 부분 최신화 했더니 생겼습니다. 혹시 이런식으로 운영을 하면 안되나요??

ksundong commented:
> > 머지를 하려고 했더니 Swift 파일이 포함되어있네요 ㅠㅠ 어떻게 된 일일까요?
> 
> @ksundong 현재 밖이라 다른 리뷰에 관해서는 추후 답변을 달겠습니다.
> 
> team-1브랜치는 ios와 be가 통합적으로 되어있는 구조입니다. 어제 ios쪽에 머지가 되어서 그 부분 최신화 했더니 생겼습니다. 혹시 이런식으로 운영을 하면 안되나요??

일단 왠만하면 별개의 프로젝트로 관리하는게 일반적이고
리뷰를 보낼 때는 자신이 작업한 것만 리뷰를 받는게 맞겠죠?

그리고, 이렇게 swift 파일이 머지가 되어버리면, ios쪽에서 리뷰를 받지 못할 수도 있고, 혹은 충돌이 발생할 수 있습니다.
